client/client_update_test.go

```diff
@@ -606,3 +636,436 @@ func testUpdateRemoteNon200Error(t *testing.T, opts updateOpts, errExpected inte
+			testUpdateRemoteCorruptValidChecksum(t, updateOpts{
+				serverHasNewData: true,
+				localCache:       true,
+				role:             role,
+			}, testData, shouldFail)
+		}
+	}
+}
+
+func testUpdateRemoteCorruptValidChecksum(t *testing.T, opts updateOpts, expt swizzleExpectations, shouldErr bool) {
+	_, serverSwizzler := newServerSwizzler(t)
+	ts := readOnlyServer(t, serverSwizzler.MetadataCache, http.StatusNotFound)
+	defer ts.Close()
+
+	repo := newBlankRepo(t, ts.URL)
+	defer os.RemoveAll(repo.baseDir)
+
+	if opts.localCache {
+		_, err := repo.Update(false)
+		require.NoError(t, err)
+	}
+
+	if opts.serverHasNewData {
+		bumpVersions(t, serverSwizzler, 1)
+	}
+
+	msg := fmt.Sprintf("swizzling %s to return: %v (forWrite: %v)", opts.role, expt.desc, opts.forWrite)
+
+	require.NoError(t, expt.swizzle(serverSwizzler, opts.role),
+		"failed %s", msg)
+
+	// update the snapshot and timestamp hashes to make sure it's not an involuntary checksum failure
+	// unless we want the server to not actually have any new data
+	if !opts.localCache || opts.serverHasNewData {
+		// we don't want to sign if we are trying to swizzle one of these roles to
+		// have a different signature - updating hashes would be pointless (because
+		// nothing else has changed) and would just overwrite the signature.
+		isSignatureSwizzle := expt.desc == "invalid signatures" || expt.desc == "meta signed by wrong key"
+		// just try to do these - if they fail (probably because they've been swizzled), that's fine
+		if opts.role != data.CanonicalSnapshotRole || !isSignatureSwizzle {
+			serverSwizzler.UpdateSnapshotHashes()
+		}
+		if opts.role != data.CanonicalTimestampRole || !isSignatureSwizzle {
+			serverSwizzler.UpdateTimestampHash()
+		}
+	}
+	_, err := repo.Update(opts.forWrite)
+	if shouldErr {
+		require.Error(t, err, "expected failure updating when %s", msg)
+
+		errType := reflect.TypeOf(err)
+		isExpectedType := false
+		var expectedTypes []string
+		for _, expectErr := range expt.expectErrs {
+			expectedType := reflect.TypeOf(expectErr)
+			isExpectedType = isExpectedType || errType == expectedType
+			expectedTypes = append(expectedTypes, expectedType.String())
+		}
+		require.True(t, isExpectedType, "expected one of %v when %s: got %s",
+			expectedTypes, msg, errType)
+
+	} else {
+		require.NoError(t, err, "expected no failure updating when %s", msg)
+	}
+}
+
+// If the local root is corrupt, and the remote root is corrupt, we should fail
+// to update.  Note - this one is really slow.
+func TestUpdateLocalAndRemoteRootCorrupt(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode")
+	}
+	for _, localExpt := range waysToMessUpLocalMetadata {
+		for _, serverExpt := range waysToMessUpServer {
+			if localExpt.desc == "expired metadata" && serverExpt.desc == "lower metadata version" {
+				// TODO: bug right now where if the local metadata is invalid, we just download a
```

Comment:
This one seems tricky. What we're saying is that: we know there is corrupt local metadata, but we're going to trust parts of that local metadata to make decisions about the validity of the remote state.

Feels like a fail open vs fail closed decision, more than a bug. Do we want the fail closed?